querytest/compare.go

```diff
@@ -0,0 +1,222 @@
+	return r.equal
+}
+func (r *results) Err() error {
+	return r.err
+}
+func (r *results) Fmt() map[string]*execute.Formatter {
+	return r.fmt
+}
+
+func (t *tables) EqualWithTolerance(s *tables, tol float64, ulp uint) *results {
+	defer t.data.Release()
+	defer s.data.Release()
+	for t.data.More() && s.data.More() {
+		l := t.data.Next()
+		r := s.data.Next()
+		if l.Name() != r.Name() {
+			return &results{err: fmt.Errorf(
+				"result names not equal - %s: %s != %s: %s",
+				t.name, l.Name(), s.name, r.Name()),
+			}
+		}
+		var lt, rt []table
+		l.Tables().Do(func(tbl flux.Table) error {
+			lt = append(lt, table{
+				name: t.name,
+				data: tbl,
+			})
+			return nil
+		})
+		r.Tables().Do(func(tbl flux.Table) error {
+			rt = append(rt, table{
+				name: s.name,
+				data: tbl,
+			})
+			return nil
+		})
+		if len(lt) != len(rt) {
+			return &results{err: fmt.Errorf(
+				"result sizes not equal - %s:%d != %s:%d",
+				t.name, len(lt), s.name, len(rt)),
+			}
+		}
+		for i, table := range lt {
+			if ok, err := table.equalWithTolerance(rt[i], tol, ulp); !ok {
+				return &results{
+					fmt: map[string]*execute.Formatter{
+						t.name: execute.NewFormatter(table.data, nil),
+						s.name: execute.NewFormatter(rt[i].data, nil),
+					},
+					err: err,
+				}
+			}
+		}
+	}
+	if t.data.More() || s.data.More() {
+		return &results{
+			err: errors.New("number of results not equal"),
+		}
+	}
+	return &results{equal: true}
+}
+
+type table struct {
+	name string
+	data flux.Table
+}
+
+func (t table) equalWithTolerance(s table, tol float64, ulp uint) (bool, error) {
+	if !equalKeys(t.data.Key(), s.data.Key()) {
+		return false, fmt.Errorf("group keys not equal - %s: %v != %s: %v",
+			t.name, t.data.Key(), s.name, s.data.Key())
+	}
+	if !equalCols(t.data.Cols(), s.data.Cols()) {
+		return false, fmt.Errorf("column schemas not equal - %s: %v != %s: %v",
+			t.name, t.data.Cols(), s.name, s.data.Cols())
+	}
+	var eq bool
+	var row int
+	return eq, t.data.Do(func(tr flux.ColReader) error {
+		return s.data.Do(func(sr flux.ColReader) error {
+			if tr.Len() != sr.Len() {
+				return fmt.Errorf("tables have different lengths - %s: %d != %s: %d",
+					t.name, tr.Len(), s.name, sr.Len())
+			}
+			for i, col := range t.data.Cols() {
+				switch col.Type {
+				case flux.TBool:
+					row, eq = equalBools(tr.Bools(i), sr.Bools(i))
+				case flux.TInt:
+					row, eq = equalInts(tr.Ints(i), sr.Ints(i))
+				case flux.TUInt:
+					row, eq = equalUInts(tr.UInts(i), sr.UInts(i))
+				case flux.TFloat:
+					row, eq = equalFloats(tr.Floats(i), sr.Floats(i), tol, ulp)
+				case flux.TString:
+					row, eq = equalStrings(tr.Strings(i), sr.Strings(i))
+				case flux.TTime:
+					row, eq = equalTimes(tr.Times(i), sr.Times(i))
+				}
+				if !eq {
+					return fmt.Errorf("values not equal at row %d column %s", row, tr.Cols()[i].Label)
+				}
+			}
+			return nil
+		})
+	})
+}
+
+func equalFloats(a, b []float64, tol float64, ulp uint) (int, bool) {
```

Comment:
Should I just add the `Equals()` method to the ColumnReader and Table interfaces?

Reply:
That is probably OK.  If you don't want to deal with the interface cascade, you can just export these so others can use them.   I'm on the fence whether we need this code outside of a testing scenario, but it might be useful some day, and odds are if we keep it in `querytest` we would forget about it and re-implement the same feature in `execute`

@nathanielc what do you think?